boltdb_iterator.go

```diff
@@ -28,33 +29,58 @@ var _ Iterator = (*boltDBIterator)(nil)
 
 // newBoltDBIterator creates a new boltDBIterator.
 func newBoltDBIterator(tx *bbolt.Tx, start, end []byte, isReverse bool) *boltDBIterator {
+	// We can check for empty key at the start, because we use a read/write transaction that blocks
+	// the entire database for writes while the iterator exists. If we change to a read-only txn
+	// that supports concurrency we'll need to rewrite this logic.
+	emptyKeyValue := tx.Bucket(bucket).Get(boltDBEmptyKey)
 	itr := tx.Bucket(bucket).Cursor()
 
 	var ck, cv []byte
 	if isReverse {
-		if end == nil {
+		switch {
+		case end == nil:
 			ck, cv = itr.Last()
-		} else {
+		case len(end) == 0:
+			// If end is the blank key, then we don't return any keys by definition
+			ck = nil
+			cv = nil
+		default:
 			_, _ = itr.Seek(end) // after key
 			ck, cv = itr.Prev()  // return to end key
 		}
+		// If we're currently positioned at the placeholder for the empty key, skip it (handle later)
+		if emptyKeyValue != nil && bytes.Equal(ck, boltDBEmptyKey) {
+			ck, cv = itr.Prev()
+		}
+		// If we didn't find any initial key, but there's a placeholder for the empty key at the
+		// end that we've skipped, then the initial key should be the empty one (the final one).
+		if emptyKeyValue != nil && ck == nil && (end == nil || len(end) > 0) {
```

Comment:
`(end == nil || len(end) > 0)` can't we rewrite this as `len(end) == 0`?

Reply:
No, what we specifically want to check is that `end` is not `[]byte{}`. `len(end) != 0` does not work, because `len(nil) == 0`.